ttnn/cpp/ttnn/operations/ccl/barrier/device/barrier_op.cpp

```diff
@@ -0,0 +1,98 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "ttnn/operations/ccl/barrier/device/barrier_op.hpp"
+#include "tt_metal/host_api.hpp"
+
+#include <cstdint>
+
+namespace ttnn {
+
+void Barrier::validate(const std::vector<Tensor>& input_tensors) const {
+    //Validate the input tensor
+    TT_FATAL(this->topology == ccl::Topology::Ring, "We currently only support Ring topologies on this OP");
+}
+
+std::vector<SimpleShape> Barrier::compute_output_shapes(const std::vector<Tensor>& input_tensors) const {
+    //For a Barrier the output shape should match the input
+    SimpleShape shape = input_tensors[0].get_logical_shape();
+    return std::vector<SimpleShape>(input_tensors.size(), shape);
+}
+
+std::vector<Tensor> Barrier::create_output_tensors(const std::vector<Tensor>& input_tensors) const {
+    //Tensor is unmodified, return what was passed in
+    return input_tensors;
+}
+
+operation::ProgramWithCallbacks Barrier::create_program(
+    const std::vector<Tensor>& input_tensors, std::vector<Tensor>& output_tensors) const {
+    return ccl::barrier::detail::barrier_with_workers(
+        input_tensors.at(0),
+        output_tensors.at(0),
+        this->is_starting_core,
+        this->ring_size,
+        this->ring_index,
+        this->receiver_device_id,
+        this->sender_device_id,
+        this->topology);
+}
+
+void Barrier::update_structure (const Tensor& input_tensor)
+{
+    //Need to resolve the neighbours of this tensor
+    //Can only be done in launch_op as it differs for different input tensors
+    const auto devices = this -> devices;
+    const bool is_linear = (topology == ttnn::ccl::Topology::Linear);
+    const uint32_t num_devices = this->ring_size;
+    uint32_t device_index = 0;
+    std::optional<chip_id_t> receiver_device_id = std::nullopt;
+    std::optional<chip_id_t> sender_device_id = std::nullopt;
+    for (uint32_t i = 0; i < num_devices; ++i) {
+        if (devices.at(i) == input_tensor.device()) {
```

Comment:
Note for future: we'll need to update this to use the factored out function that @Aswinmcw is pulling out so this device selection logic is using the commonized implementation.